### PR TITLE
Feature/sg 635 fix dataloader multiprocessing

### PR DESCRIPTION
**Pull request by Louis-Dupont**
Remaining: having Eugene and Tomer try:
- num_workers = 0 and num_workers = 8
- Without DDP and With DDP (gpus>1)
Can be done on Cifar10 for 2 batches, no need to work with lots of data and/or run for w while.

Expected: See the following message as many times as num_workers
`FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=`

### CODE REVIEW

File: src/super_gradients/common/abstractions/mute_processes.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        Knowing that depending on how the script is launched, main_process might be child of other non "python" processes such as:
+                ssh(non-python) -> pycharm(non-python) -> main_process(python) -> ...
+    """
+
+    if any(os_name in platform.platform() for os_name in ["macOS", "Windows"]):
+
+        # When using DDP, we expect the worker process to have 2 parents processes using python, and only 1 otherwise.
+        # Note that this is a "root_process" is the root process only if current process is a worker process
+        main_process = psutil.Process().parent() if int(env_variables.LOCAL_RANK) == -1 else psutil.Process().parent().parent()
+        is_worker_process = main_process and "python" in main_process.name()
```
Comment (BloodAxe) [inline src/super_gradients/common/abstractions/mute_processes.py:63]: What if user starts DDP with `torchrun`? Would main process still contain `python` name? 
Perhaps, it is less relevant for Windows & macOS as DDP is not that popular/possible in these platforms.

**Comment by BloodAxe** [comment]
This is the output I've got when running on Windows

```
C:\Anaconda3\envs\sg\python.exe D:\Develop\GitHub\Deci\super-gradients\src\super_gradients\examples\train_from_recipe_example\train_from_recipe.py --config-name=cifar10_resnet dataset_params.train_dataloader_params.num_workers=4 dataset_params.val_dataloader_params.num_workers=4 multi_gpu=Off num_gpus=1 
The console stream is logged into C:\Users\blood\sg_logs\console.log
[2023-03-06 12:48:24] INFO - crash_tips_setup.py - Crash tips is enabled. You can set your environment variable to CRASH_HANDLER=FALSE to disable it
[2023-03-06 12:48:25] WARNING - redirects.py - NOTE: Redirects are currently not supported in Windows or MacOs.
C:\Anaconda3\envs\sg\lib\site-packages\_distutils_hack\__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
[2023-03-06 12:48:29] WARNING - __init__.py - Failed to import pytorch_quantization
[2023-03-06 12:48:29] WARNING - calibrator.py - Failed to import pytorch_quantization
[2023-03-06 12:48:29] WARNING - export.py - Failed to import pytorch_quantization
[2023-03-06 12:48:29] WARNING - selective_quantization_utils.py - Failed to import pytorch_quantization
[2023-03-06 12:48:29] WARNING - env_sanity_check.py - Failed to verify operating system: Deci officially supports only Linux kernels. Some features may not work as expected.
[2023-03-06 12:48:29] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:48:29] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
[2023-03-06 12:48:29,807][super_gradients.training.models.model_factory][WARNING] - Passing num_classes through arch_params is deprecated and will be removed in the next version. Pass num_classes explicitly to models.get
Downloading https://www.cs.toronto.edu/~kriz/cifar-10-python.tar.gz to ./data/cifar10\cifar-10-python.tar.gz
100%|██████████| 170498071/170498071 [00:19<00:00, 8876408.50it/s]
Extracting ./data/cifar10\cifar-10-python.tar.gz to ./data/cifar10
Files already downloaded and verified
[2023-03-06 12:48:54,088][super_gradients.training.sg_trainer.sg_trainer][WARNING] - Train dataset size % batch_size != 0 and drop_last=False, this might result in smaller last batch.
The console stream is now moved to D:\Develop\GitHub\Deci\super-gradients\checkpoints\resnet18_cifar/console_март06_12_48_54.txt
FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=Windows-10-10.0.19041-SP0
[2023-03-06 12:49:03] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:49:03] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=Windows-10-10.0.19041-SP0
[2023-03-06 12:49:13] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:49:13] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=Windows-10-10.0.19041-SP0
[2023-03-06 12:49:21] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:49:21] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=Windows-10-10.0.19041-SP0
[2023-03-06 12:49:29] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:49:29] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
[2023-03-06 12:49:31] INFO - sg_trainer_utils.py - TRAINING PARAMETERS:
    - Mode:                         OFF
    - Number of GPUs:               1          (1 available on the machine)
    - Dataset size:                 50000      (len(train_set))
    - Batch size per GPU:           256        (batch_size)
    - Batch Accumulate:             1          (batch_accumulate)
    - Total batch size:             256        (num_gpus * batch_size)
    - Effective Batch size:         256        (num_gpus * batch_size * batch_accumulate)
    - Iterations per epoch:         195        (len(train_set) / total_batch_size)
    - Gradient updates per epoch:   195        (len(train_set) / effective_batch_size)

[2023-03-06 12:49:31] INFO - sg_trainer.py - Started training for 250 epochs (0/249)

Train epoch 0:   0%|          | 0/196 [00:00<?, ?it/s]FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=Windows-10-10.0.19041-SP0
[2023-03-06 12:49:39] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:49:39] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=Windows-10-10.0.19041-SP0
[2023-03-06 12:49:49] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:49:49] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=Windows-10-10.0.19041-SP0
[2023-03-06 12:49:58] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:49:58] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
FOR DEBUGGING ONLY: Current process is a worker process. Muting dataloader workers. platform=Windows-10-10.0.19041-SP0
[2023-03-06 12:50:07] ERROR - env_sanity_check.py - Failed to verify installed packages: pip-tools==6.9.0 does not satisfy requirement pip-tools>=6.12.1
[2023-03-06 12:50:07] ERROR - env_sanity_check.py - Failed to verify installed packages: onnx-simplifier required but not found
```

The bottom line - the logic to discriminate whether we're inside worker processes works.
However it is still emiting error messages from env_sanity_check() which IMO is the root cause of the problem - it should be called once in main() and not on import level.

File: src/super_gradients/common/abstractions/mute_processes.py (hunk 1 above)
Comment (Louis-Dupont) [reply src/super_gradients/common/abstractions/mute_processes.py:78]: In the case of someone using torch run on Windows/macOS we will still see all the logs like before. 
I pushed a change that should handle it but I cannot test DDP on mac so it's hard to say if it works or not.

https://github.com/Deci-AI/super-gradients/blob/9c03e1b9ba9cf7724fef2208edeb4492ae62fc7f/src/super_gradients/common/abstractions/mute_processes.py#L68-L76

**Comment by Louis-Dupont** [comment]
I removed the "`FOR DEBUGGING ONLY: ... `" print and also made so that env_sanity_check is not run on subprocesses

**Review by BloodAxe — APPROVED** [review]
LGTM
